src/ucp/core/ucp_context.c

```diff
@@ -865,62 +859,103 @@ static int ucp_is_resource_in_device_list(const uct_tl_resource_desc_t *resource
-        snprintf(info, sizeof(info), " for alias '%s'", alias->alias);
-        dummy_mask      = 0;
-        tmp_rsc_flags   = 0;
         tmp_tl_cfg_mask = 0;
-        dummy_flags     = 0;
-        if (ucp_tls_array_is_present(alias->tls, alias_arr_count, tl_name, info,
-                                     &tmp_rsc_flags, &dummy_mask)) {
-            if (ucp_config_is_tl_name_present(allow_list, alias->alias, 1,
-                                              &dummy_flags, &tmp_tl_cfg_mask)) {
-                if (allow_list->mode == UCS_CONFIG_ALLOW_LIST_ALLOW) {
-                    *rsc_flags   |= tmp_rsc_flags;
-                    *tl_cfg_mask |= tmp_tl_cfg_mask;
-                    return 1;
-                } else if (!(tmp_rsc_flags & UCP_TL_RSC_FLAG_AUX)) {
-                    /* Disable the transport only if it is not used in the alias
-                     * as an auxiliary transport. */
-                    return 0;
-                }
+        if (!ucp_config_is_tl_name_present(tl_array, alias->alias, 1, check_aux,
+                                           &tmp_tl_cfg_mask)) {
+            continue;
+        }
+
+        alias_arr_count = ucp_tl_alias_count(alias);
+        if (ucp_tls_array_is_present(alias->tls, alias_arr_count, tl_name, 0,
+                                     &dummy_mask)) {
+            *tl_cfg_mask |= tmp_tl_cfg_mask;
+            return is_allow_mode;
+        }
+
+        /* Only transports mentioned without 'aux' suffix in the alias are
+         * considered in case of deny list. E.g:
+         * alias = { tl_name_1, tl_name_2:aux }, UCX_TLS=^...alias,... should
+         * not disable tl_name_2. */
+        if (is_allow_mode) {
+            if (ucp_tls_array_is_present(alias->tls, alias_arr_count, tl_name,
+                                         1, &dummy_mask)) {
+                *tl_cfg_mask |= tmp_tl_cfg_mask;
+                enable_aux    = 1;
             }
         }
     }
 
-    return allow_list->mode == UCS_CONFIG_ALLOW_LIST_NEGATE;
+    if (enable_aux) {
+        *rsc_flags |= UCP_TL_RSC_FLAG_AUX;
+        return 1;
+    }
+
+    return !is_allow_mode;
+}
+
+static int
+ucp_is_resource_in_transports_list(const char *tl_name,
+                                   const ucs_config_allow_list_t *allow_list,
+                                   const ucs_string_set_t *aux_tls,
+                                   uint8_t *rsc_flags, uint64_t *tl_cfg_mask)
+{
+    const int is_allow_mode = (allow_list->mode == UCS_CONFIG_ALLOW_LIST_ALLOW);
+    int enable;
+
+    if (allow_list->mode == UCS_CONFIG_ALLOW_LIST_ALLOW_ALL) {
+        return 1;
+    }
+
+    ucs_assert(allow_list->array.count > 0);
+    enable = ucp_is_tl_in_transports_list(tl_name, &allow_list->array,
+                                          is_allow_mode, 0, rsc_flags,
+                                          tl_cfg_mask);
+
+    if (is_allow_mode || !ucs_string_set_contains(aux_tls, tl_name) || enable) {
```

Comment:
seems weird, I would expect something like
if (enable) return is_allow_mode;
can you pls document the logic?